### ISSUE TITLE
Update CliMT initialization handling and component dimensions

### ISSUE THREAD

**Issue by mcgibbon** [CliMT/climt#30]
I'm proposing the following changes to the way properties dictionaries and initialization are configured on CliMT, which I can hopefully implement in a fairly short period of time.

Components:

- require specific dimension names if they care about that dimension in some way, for example the dynamical core likely cares what the horizontal and vertical dimensions are, and will have no wildcards. The dimension names should be static (not customizable), so that they are guaranteed to be properly documenting.
- use `"*"` to collect any dimensions they do not care about. For example, a column parameterization would for a mid-level variable require dims to be `["*", "mid_levels"]`. SimplePhysics should likely be like this, since it only needs latitude as a variable and doesn't care about the configuration of the latitudinal grid (as I understand it).
- make it so the default grid specification from get_default_state should not result in array copying on input when collecting dimensions using `["*", "mid_levels"]`. This is pretty easy to achieve.
- specify their properties dictionaries explicitly, potentially with the use of helper functions to reduce boilerplate code. I will figure out those helper functions by initially writing out the properties for a component manually, and thinking about how any unnecessary boilerplate could be reduced. I will look at the existing functions that do this for reference.
- check arrays for contiguity before passing into any Fortran code, and copy the data to a contiguous array if necessary using `np.ascontiguousarray`. The default array configuration should not result in copying after collecting dimensions using `["*", "mid_levels"]` (this should also be pretty easy to achieve).

Initialization:

- should have a simple mode as it currently does where you just pass in a set of components and get a reasonable global state.
- should have an additional mode where you use one function to get a "grid" state (that sets longitude and latitude with the number of dimensions you want), and that "grid" state can be used to get the full state. This could look something like:
```
grid_state = get_grid(*args, **kwargs)
state = get_default_state(component_list, grid_state=grid_state)
```
- This mode would allow users to manually specify their horizontal and vertical grids, but still get the utility of automatic default variables. The grid state could have only a vertical dimension with longitude and latitude as 0-dimensional scalars if a user wants a single-column model.

Ideally I can implement all of this in about a week for @JoyMonteiro to review, including the changes necessary on all components.

**Comment by JoyMonteiro** [comment]
@mcgibbon why don't you go ahead and create a fork of develop (or a new branch) and implement these changes. Once they look good, we could merge them into develop. I will finish merging the PR by then. You are correct in saying that SimplePhysics does not care about grid configuration, and a flattened `latitude` array with x*y elements will work just fine.

I recently pushed some changes to `ArrayHandler` which automate creation of properties dictionaries. That should hopefully provide a reference.

As a first cut, it would be good to only modify those components required for
https://raw.githubusercontent.com/CliMT/climt/develop/examples/full_radiation_with_insolation_gcm.py

and get this example to successfully work (with openMP). As you will notice, this will involve modifications all the way down to the cython level. There are some memory copies which are killing performance, and it would be great if these changes are able to get rid of those.

I'm fine with the proposed `get_grid` functionality. However, looking ahead to `Federation`, it would be good if `get_grid` can be used automatically -- it takes in a list of components, checks their `grid_description` attribute (if any) and returns a grid_state compatible with them.

This way, in `Federation` I can call `get_grid` always and get back a grid compatible with a single-column model if none of the component have a `grid_description`.

**Comment by JoyMonteiro** [comment]
I have been thinking a little more about this, and I'm not convinced that dimension names should be static. Let's say you have a general purpose fluid dynamics solver which can work given grid spacing and grid points. It could be used, for example, as a coastal ocean model or a limited area atmosphere model.

Then, there is no way to know in advance what the dimension names will be, and it makes sense that these dimension names be meaningful and should be set in calls to `get_default_state`.

My proposal is that `*_properties` be created at runtime, as it currently is happening, and the 'x' and 'y' dimensions be replaced with appropriate dimension names during runtime.

 'x', 'y' are properly documented dimensions when we don't know what they will be in advance, and they will be replaced with the correct names when these names are known.

**Comment by mcgibbon** [comment]
I'll keep that in mind and definitely make sure there's a simple get-state-from-list-of-components mode, whether that involves get_grid we'll see. In my eyes get_grid is a way for the user to manually specify properties of the grid, and get_default_state can handle getting a default grid for components if none exists already.

*_properties actually has to be created on initialization rather than at runtime for compatibility with Sympl 0.4.0. The components that are single-column have no actual reason to require x and y horizontal dimensions - they only care about the column number. It is overly restrictive to them to require those dimensions. In fact, under how Sympl is currently written for 0.4.0, this would mean those components would not be able to operate in a single-column mode unless at least one of their inputs had an explicit "x" or "y" dimension (since it needs to be able to determine the dimension length). This can be changed, but I'll have to think about whether it should be, since this would mean for example that a model of strictly 2D inputs could be used as input for a `['*', 'mid_levels']`, `['*', 'interface_levels]` component, and I'm not sure yet whether that's desirable.

The purpose of wildcards in Sympl is to handle unknown dimension names, and this case of unknown horizontal dimension names can be handled by the `'*'` wildcard.

If you do have a general purpose solver that can work given an arbitrary irregular grid, then you would use ['*', <vertical_dimension>]. But you don't, right? The current dynamics solver requires constant-latitude and constant-longitude axes, if I remember correctly. In the future, it would be very easy to implement changes allowing an arbitrary lat-lon configuration if we have a dynamics solver that can make use of it.

**Comment by mcgibbon** [comment]
When I say that 'x' and 'y' are not properly documented dimensions for a column-wise component, what I mean is that they imply the component actually wants a horizontal grid, which it in reality doesn't care about. And for the dynamical core, they are just the incorrect dimension names, in the case that it requires a constant-latitude, constant-longitude grid.

**Comment by JoyMonteiro** [comment]
Why would *_properties be required at initialisation? AFAIU, *_properties is used by the array handling code to create compatible arrays, and this information is not going to be used until after the component has been initialised. If you mean it has to be available after initialisation, then the current way of making *_properties an  object property (pardon the confusing terminology) should work.

Is there an actual use case for a component which can only work in single column mode, and will fail if given multiple columns to iterate over?

My concern is not for the way the solver will function, nor if the grid is irregular or not. It certainly won't work if the horizontal dimensions are collapsed (How will it know what is x and y to calculate gradients, for example, or it will need to reshape dimensions when this is not required). 

My concern is for how the output will be formatted, since dimension names are really only important there. Right now, I can name dimensions as 'eastward_from_coast' or as 'latitude', and the output will make sense. If we don't allow any way of providing dimension names, then all output will be cryptically labeled 'x', 'y' or some generic label specified in the solver which is not desirable.

**Comment by mcgibbon** [comment]
In short because some wrappers need to modify the properties (particularly tendencies_in_diagnostics), which cannot be done with a property function.

A component that is single column with a `'*'` wildcard for its column dimensions will not fail when given multiple columns to iterate over.

The dynamics solver (that we currently have) would not collapse the dimensions using `'*'`, it would explicitly specify them (since it actually needs the grid to conform to a certain specification).

The output from the components will have the same dimension names as the input of the components. `'*`` automatically restores the original column dimensions when returning DataArrays. And when using the dynamics solver, it wouldn't be a cryptic dimension you would be specifying - the dimension name should get across that dimension's actual purpose in the model grid (probably `lat` and `lon`).

**Comment by JoyMonteiro** [comment]
In the case of that hypothetical general purpose solver, 'x' and 'y' are the correct dimension names, since they could represent anything. They have to be overridden by the model script to have a meaningful name depending on the context.

Even in the case of the dycore, I could choose to name the dimensions 'degrees_east_of_seattle' and 'degrees_north_of_cape_town', which are as meaningful as 'latitude', 'longitude'. I still don't see why we should not allow the user to provide names of dimensions.

**Comment by JoyMonteiro** [comment]
why should a wrapper modify the *_properties of an input component? The way I see it, it should create a new *_properties based on the *_properties of the input component. It seems like a bad idea to modify the *_properties of a component...

**Comment by mcgibbon** [comment]
It isn't a wrapper doing the modification, it's the base class, and it's done on initialization only. It's mainly to put tendency properties in diagnostics when using that option.

**Comment by mcgibbon** [comment]
Going back to the discussion, if we have a hypothetical general purpose solver, then we could set its dimension names appropriately when we get it. This restriction would only be made by the dycore, and get_default_state would be written to accommodate that restriction.

'degrees_east_of_seattle' may be just as meaningful as 'longitude', but just the same you could say 'temperature_of_the_air' is just as meaningful as 'air_temperature', and as a result we need a way to allow the user to specify quantity names. In practice, it's sufficient to set one of them in the model, and users can modify quantity names and dimension names and whatever other names in any way they want before they output/plot those variables. The reason we shouldn't have the user change model dimension names is the same as the reason we shouldn't have them change model variable names - because it's not necessary, there is no use case it enables that isn't already enabled by fairly simple workarounds, and it leads to a lot of code bloat.

If a user really wants this functionality, they should write a function `alias_state(state)` which takes in the model state and modifies dimension and variable names in the way they like, and they should call that on state before using it in any non-CliMT code.

I'm ready to move forward with these changes if you're convinced.

**Comment by JoyMonteiro** [comment]
In the very near future, I want to integrate an LES/CRM into climt:
http://www2.mmm.ucar.edu/people/bryan/cm1/

This has a grid defined in meters. So, I need a way to specify what it is that the
grid corresponds to. If we make it so that the user cannot change the names
of the dimensions, then the dimension names have to be something generic
like 'x_dimension' or something. Since my main concern is what the fields get
named in the output, then we could extend the alias functionality in netcdf monitor
to rename the dimensions while writing to disk.

Does this sound better?

**Comment by mcgibbon** [comment]
Yeah, that sounds fine. The dynamical core for the LES/CRM would just have the generic x/y dimension name we want set in its properties. The alias functionality for dimensions isn't currently written, but can easily be added.

**Comment by mcgibbon** [comment]
Fixed by #68
